pkg/eks/podidentities/procedures.go

```diff
@@ -0,0 +1,102 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podidentities
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/aws/aws-sdk-go/service/eks/eksiface"
+)
+
+var (
+	// ErrNilPodIdentityAssociation defines an error for when a nil eks pod identity is returned.
+	ErrNilPodIdentityAssociation = errors.New("nil eks pod identity association returned from create")
+	// ErrPodIdentityAssociationNotFound defines an error for when an eks pod identity is not found.
+	ErrPodIdentityAssociationNotFound = errors.New("eks pod identity association not found")
+	// ErrPodIdentityAssociationAlreadyExists defines an error for when an eks pod identity already exists.
+	ErrPodIdentityAssociationAlreadyExists = errors.New("eks pod identity association already exists")
+)
+
+// DeletePodIdentityAssociationProcedure is a procedure that will delete an EKS eks pod identity.
+type DeletePodIdentityAssociationProcedure struct {
+	eksClient             eksiface.EKSAPI
+	clusterName           string
+	existingAssociationID string
+}
+
+// Do implements the logic for the procedure.
+func (p *DeletePodIdentityAssociationProcedure) Do(_ context.Context) error {
+	input := &eks.DeletePodIdentityAssociationInput{
+		AssociationId: aws.String(p.existingAssociationID),
+		ClusterName:   aws.String(p.clusterName),
+	}
+
+	if _, err := p.eksClient.DeletePodIdentityAssociation(input); err != nil {
+		return fmt.Errorf("deleting eks pod identity %s: %w", p.existingAssociationID, err)
+	}
+
+	return nil
+}
+
+// Name is the name of the procedure.
+func (p *DeletePodIdentityAssociationProcedure) Name() string {
+	return "eks_pod_identity_delete"
+}
+
+// CreatePodIdentityAssociationProcedure is a procedure that will create an EKS eks pod identity for a cluster.
+type CreatePodIdentityAssociationProcedure struct {
+	eksClient      eksiface.EKSAPI
+	clusterName    string
+	newAssociation *EKSPodIdentityAssociation
+}
+
+// Do implements the logic for the procedure.
+func (p *CreatePodIdentityAssociationProcedure) Do(_ context.Context) error {
+	if p.newAssociation == nil {
+		return fmt.Errorf("getting desired eks pod identity for cluster %s: %w", p.clusterName, ErrPodIdentityAssociationNotFound)
+	}
+
+	input := &eks.CreatePodIdentityAssociationInput{
+		ClusterName:    aws.String(p.clusterName),
+		Namespace:      p.newAssociation.ServiceAccountNamespace,
+		RoleArn:        p.newAssociation.RoleARN,
+		ServiceAccount: p.newAssociation.ServiceAccountName,
+	}
+
+	output, err := p.eksClient.CreatePodIdentityAssociation(input)
+	if err != nil {
+		return fmt.Errorf("creating desired eks pod identity for cluster %s: %w", p.clusterName, err)
+	}
+
+	if output == nil {
+		return ErrNilPodIdentityAssociation
+	}
+	if output.Association == nil {
+		return ErrNilPodIdentityAssociation
+	}
```

Comment:
Can this really happen? Or is there a specific reason why we want to check this field here, since we're not even reading it? I'd rather remove this.

Reply:
If there is a possibility for it in their SDK, I personally wouldnt rule it out. I'd personally rather catch the possibility of this error loudly with an error returned, than letting it silently continue. Ill follow your lead and remove it as the arn is not read/used at present